// vscode-languageclient/lib/selectionRange.proposed.js
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const UUID = require("./utils/uuid");
const Is = require("./utils/is");
const vscode_1 = require("vscode");
const vscode_languageserver_protocol_1 = require("vscode-languageserver-protocol");
const client_1 = require("./client");
function ensure(target, key) {
    if (target[key] === void 0) {
        target[key] = {};
    }
    return target[key];
}
class SelectionRangeFeature extends client_1.TextDocumentFeature {
    constructor(client) {
        super(client, vscode_languageserver_protocol_1.SelectionRangeRequest.type);
    }
    fillClientCapabilities(capabilites) {
        let capability = ensure(ensure(capabilites, 'textDocument'), 'selectionRange');
        capability.dynamicRegistration = true;
    }
    initialize(capabilities, documentSelector) {
        if (!capabilities.selectionRangeProvider) {
            return;
        }
        const implCapabilities = capabilities.selectionRangeProvider;
        const id = Is.string(implCapabilities.id) && implCapabilities.id.length > 0 ? implCapabilities.id : UUID.generateUuid();
        const selector = implCapabilities.documentSelector || documentSelector;
        if (selector) {
            this.register(this.messages, {
                id,
                registerOptions: Object.assign({}, { documentSelector: selector })
            });
        }
    }
    registerLanguageProvider(options) {
        let client = this._client;
        let provideSelectionRanges = (document, positions, token) => {
            const requestParams = {
                textDocument: client.code2ProtocolConverter.asTextDocumentIdentifier(document),
                positions: client.code2ProtocolConverter.asPositions(positions)
            };
            return client.sendRequest(vscode_languageserver_protocol_1.SelectionRangeRequest.type, requestParams, token).then((ranges) => this.asSelectionRanges(ranges), (error) => {
                client.logFailedRequest(vscode_languageserver_protocol_1.SelectionRangeRequest.type, error);
                return Promise.resolve(null);
            });
        };
        let middleware = client.clientOptions.middleware;
        return vscode_1.languages.registerSelectionRangeProvider(options.documentSelector, {
            provideSelectionRanges(document, positions, token) {
                return middleware.provideSelectionRanges
                    ? middleware.provideSelectionRanges(document, positions, token, provideSelectionRanges)
                    : provideSelectionRanges(document, positions, token);
            }
        });
    }
    asSelectionRanges(selectionRanges) {
        if (!Array.isArray(selectionRanges)) {
            return [];
        }
        let result = [];
        for (let ranges of selectionRanges) {
            if (!Array.isArray(ranges)) {
                return [];
            }
            let inner = [];
            result.push(inner);
            for (let range of ranges) {
                inner.push(new vscode_1.SelectionRange(this._client.protocol2CodeConverter.asRange(range.range), this.asSelectionRangeKind(range.kind)));
            }
        }
        return result;
    }
    asSelectionRangeKind(kind) {
        switch (kind) {
            case vscode_languageserver_protocol_1.SelectionRangeKind.Empty:
                return vscode_1.SelectionRangeKind.Empty;
            case vscode_languageserver_protocol_1.SelectionRangeKind.Statement:
                return vscode_1.SelectionRangeKind.Statement;
            case vscode_languageserver_protocol_1.SelectionRangeKind.Declaration:
                return vscode_1.SelectionRangeKind.Declaration;
            default:
                return vscode_1.SelectionRangeKind.Empty;
        }
    }
}
exports.SelectionRangeFeature = SelectionRangeFeature;
